### ISSUE TITLE
BufferUnderflowException in CachedReadRequestChain

### ISSUE THREAD

**Issue by stagraqubole** [qubole/rubix#371]
Saw this while testing https://github.com/qubole/rubix/pull/368. One query failed with " Malformed ORC file." error and around same time another query's CachedReadRequestChain saw BufferUnderflowException but it recovered from it by fallback to remote and didnt fail.

Logs:
```
2020-05-04T12:33:34.537Z        ERROR   SplitRunner-119-593     io.prestosql.execution.executor.TaskExecutor    Error processing Split 20200504_122715_00206_sv9bs.1.0-1024 {path=s3://qubole-pres/stagra/tables/tpcds/orc/store_sales_100_sortedOnCustomer_exploded/file5, start=7918845952, length=134217728, fileSize=13877966505, hosts=[10.0.1.155], database=default, table=store_sales_100_orc_sorted_oncustomer_exploded15x, forceLocalScheduling=false, partitionName=<UNPARTITIONED>, s3SelectPushdownEnabled=false} (start = 9475486.088059, wall = 78 ms, cpu = 0 ms, wait = 0 ms, calls = 1): HIVE_BAD_DATA: Malformed ORC file. Invalid stripe footer [s3://qubole-pres/stagra/tables/tpcds/orc/store_sales_100_sortedOnCustomer_exploded/file5]
2020-05-04T12:33:34.537Z        DEBUG   20200504_122715_00206_sv9bs.1.0-1022-78 com.qubole.rubix.core.ReadRequestChain  Request to add ReadRequest: [8246001664, 8247050240, 8246845625, 8247050240, 0]
2020-05-04T12:33:34.537Z        DEBUG   20200504_123139_00220_sv9bs.1.0-294-74  com.qubole.rubix.core.ReadRequestChain  Updated last to: [1000341504, 1003487232, 1001253607, 1003487232, 0]
2020-05-04T12:33:34.537Z        DEBUG   20200504_122715_00206_sv9bs.1.0-1022-78 com.qubole.rubix.core.ReadRequestChain  Added ReadRequest: [8246001664, 8247050240, 8246845625, 8247050240, 0]
2020-05-04T12:33:34.538Z        DEBUG   20200504_122715_00206_sv9bs.1.0-1030-85 com.qubole.rubix.core.CachedReadRequestChain    Processing readrequest 13877950121-13877966505, length 16384
2020-05-04T12:33:34.538Z        DEBUG   20200504_123139_00220_sv9bs.1.0-294-74  com.qubole.rubix.core.CachingInputStream        Sending cached block 957 to cachedReadRequestChain
2020-05-04T12:33:34.538Z        DEBUG   20200504_122715_00206_sv9bs.1.0-1030-85 com.qubole.rubix.core.CachedReadRequestChain    CachedFileRead copied data [13877950121 - 13877966505] at buffer offset 0
2020-05-04T12:33:34.538Z        DEBUG   20200504_123139_00220_sv9bs.1.0-295-76  com.qubole.rubix.core.CachingInputStream        Sending cached block 1909 to cachedReadRequestChain
2020-05-04T12:33:34.538Z        DEBUG   20200504_122715_00206_sv9bs.1.0-1030-85 com.qubole.rubix.core.CachedReadRequestChain    Read 16384 bytes from cached file
2020-05-04T12:33:34.538Z        DEBUG   20200504_122715_00206_sv9bs.1.0-1022-78 com.qubole.rubix.core.CachingInputStream        Sending cached block 7865 to cachedReadRequestChain
2020-05-04T12:33:34.538Z        DEBUG   20200504_123139_00220_sv9bs.1.0-294-74  com.qubole.rubix.core.ReadRequestChain  Request to add ReadRequest: [1003487232, 1004535808, 1003487232, 1004535808, 2233625]
2020-05-04T12:33:34.538Z        DEBUG   20200504_122715_00206_sv9bs.1.0-1030-85 com.qubole.rubix.core.CachedReadRequestChain    Read Request threadName: 20200504_123139_00220_sv9bs.1.0-291-88, Cached read Executor threadName: 20200504_122715_00206_sv9bs.1.0-1030-85
2020-05-04T12:33:34.538Z        DEBUG   20200504_122715_00206_sv9bs.1.0-1030-85 com.qubole.rubix.core.CachingInputStream        Read 16384 bytes
2020-05-04T12:33:34.538Z        DEBUG   20200504_123139_00220_sv9bs.1.0-295-76  com.qubole.rubix.core.ReadRequestChain  Request to add ReadRequest: [2001731584, 2002780160, 2002290953, 2002780160, 0]
2020-05-04T12:33:34.538Z        DEBUG   20200504_122715_00206_sv9bs.1.0-1022-78 com.qubole.rubix.core.ReadRequestChain  Request to add ReadRequest: [8247050240, 8248098816, 8247050240, 8248098816, 204615]
2020-05-04T12:33:34.538Z        DEBUG   20200504_123139_00220_sv9bs.1.0-294-74  com.qubole.rubix.core.ReadRequestChain  Updated last to: [1000341504, 1004535808, 1001253607, 1004535808, 0]
2020-05-04T12:33:34.539Z        DEBUG   20200504_122715_00206_sv9bs.1.0-1030-85 com.qubole.rubix.core.CachingInputStream        New nextReadPosition: 13877966505 nextReadBlock: 13235
2020-05-04T12:33:34.539Z        DEBUG   20200504_123139_00220_sv9bs.1.0-291-88  com.qubole.rubix.core.CachedReadRequestChain    Processing readrequest 13877950121-13877966505, length 16384
2020-05-04T12:33:34.539Z        DEBUG   20200504_122715_00206_sv9bs.1.0-1022-78 com.qubole.rubix.core.ReadRequestChain  Updated last to: [8246001664, 8248098816, 8246845625, 8248098816, 0]
2020-05-04T12:33:34.539Z        DEBUG   20200504_123139_00220_sv9bs.1.0-295-76  com.qubole.rubix.core.ReadRequestChain  Added ReadRequest: [2001731584, 2002780160, 2002290953, 2002780160, 0]
2020-05-04T12:33:34.539Z        DEBUG   20200504_123139_00220_sv9bs.1.0-294-74  com.qubole.rubix.core.CachingInputStream        Sending cached block 958 to cachedReadRequestChain
2020-05-04T12:33:34.539Z        DEBUG   20200504_123139_00220_sv9bs.1.0-291-88  com.qubole.rubix.core.CachedReadRequestChain    CachedFileRead copied data [13877950121 - 13877966505] at buffer offset 0
2020-05-04T12:33:34.539Z        DEBUG   20200504_123139_00220_sv9bs.1.0-291-88  com.qubole.rubix.core.CachedReadRequestChain    Read 16384 bytes from cached file
2020-05-04T12:33:34.538Z        ERROR   20200504_123139_00220_sv9bs.1.0-293-90  com.qubole.rubix.core.CachedReadRequestChain    Fall back to read from object store for /media/ephemeral1/fcache/qubole-pres/stagra/tables/tpcds/orc/store_sales_100_sortedOnCustomer_exploded/file10 .Could not read data from cached file :
java.nio.BufferUnderflowException
        at java.base/java.nio.DirectByteBuffer.get(DirectByteBuffer.java:295)
        at com.qubole.rubix.core.CachedReadRequestChain.call(CachedReadRequestChain.java:116)
        at com.qubole.rubix.core.CachedReadRequestChain.call(CachedReadRequestChain.java:37)
        at com.google.shaded.shaded.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:125)
        at com.google.shaded.shaded.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:57)
        at com.google.shaded.shaded.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:834)
```

**Comment by stagraqubole** [comment]
BufferUnderflowException ideally shouldnt happen as we only try to read from ByteBuffer as much as we wrote to it. But this happening almost in parallel with  " Malformed ORC file." which came by reading invalid data from another CachedReadRequestChain, it looks like the buffer from BufferPool was shared between these two chains.

**Comment by stagraqubole** [comment]
There were query cancellations happening in parallel to running queries. The more probable cause of the issue is that CachingInputStream returned the buffer to pool due to the cancellation while CachedReadRequestChain was still executing in the readService thread.

**Comment by stagraqubole** [comment]
It was the case of buffer returned to pool while the were still with CachedRRC. Fixing it via https://github.com/qubole/rubix/pull/368, dont see the error with the fix.

**Comment by stagraqubole** [comment]
Fixed by Fix parallel warmup #368
